File: Graph/ts/bestBridge.ts
import { Grid, inBounds, posKey } from "./grid";

export function bestBridge(grid: Grid): number {
  const firstIsland = new Set<string>();
  outer: for (let r = 0; r < grid.length; r += 1) {
    for (let c = 0; c < grid[0].length; c += 1) {
      if (grid[r][c] === "L") {
        collectIsland(grid, r, c, firstIsland);
        if (firstIsland.size) {
          break outer;
        }
      }
    }
  }

  if (!firstIsland.size) return 0;

  const queue: Array<[number, number, number]> = [];
  const visited = new Set<string>(firstIsland);

  for (const key of firstIsland) {
    const [row, col] = key.split(",").map(Number);
    queue.push([row, col, 0]);
  }

  const deltas = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
  ];

  while (queue.length > 0) {
    const [row, col, dist] = queue.shift()!;
    const key = posKey(row, col);
    if (grid[row][col] === "L" && !firstIsland.has(key)) {
      return dist - 1;
    }

    for (const [dr, dc] of deltas) {
      const nr = row + dr;
      const nc = col + dc;
      const neighborKey = posKey(nr, nc);
      if (!inBounds(grid, nr, nc) || visited.has(neighborKey)) continue;
      visited.add(neighborKey);
      queue.push([nr, nc, dist + 1]);
    }
  }

  return -1;
}

function collectIsland(grid: Grid, r: number, c: number, island: Set<string>): void {
  if (!inBounds(grid, r, c)) return;
  if (grid[r][c] === "W") return;

  const key = posKey(r, c);
  if (island.has(key)) return;
  island.add(key);

  collectIsland(grid, r - 1, c, island);
  collectIsland(grid, r + 1, c, island);
  collectIsland(grid, r, c - 1, island);
  collectIsland(grid, r, c + 1, island);
}

File: Graph/python/bestBridge.py
from collections import deque
from typing import Deque

from .graph_utils import Grid, in_bounds, pos_key


def best_bridge(grid: Grid) -> int:
    first_island: set[str] = set()

    for r in range(len(grid)):
        for c in range(len(grid[0])):
            if grid[r][c] == "L":
                _collect_island(grid, r, c, first_island)
                if first_island:
                    break
        if first_island:
            break

    if not first_island:
        return 0

    queue: Deque[tuple[int, int, int]] = deque()
    visited = set(first_island)
    for key in first_island:
        row, col = map(int, key.split(","))
        queue.append((row, col, 0))

    deltas = [(-1, 0), (1, 0), (0, -1), (0, 1)]

    while queue:
        row, col, dist = queue.popleft()
        key = pos_key(row, col)
        if grid[row][col] == "L" and key not in first_island:
            return dist - 1

        for dr, dc in deltas:
            nr, nc = row + dr, col + dc
            nkey = pos_key(nr, nc)
            if not in_bounds(grid, nr, nc) or nkey in visited:
                continue
            visited.add(nkey)
            queue.append((nr, nc, dist + 1))

    return -1


def _collect_island(grid: Grid, r: int, c: int, island: set[str]) -> None:
    if not in_bounds(grid, r, c) or grid[r][c] == "W":
        return
    key = pos_key(r, c)
    if key in island:
        return
    island.add(key)

    _collect_island(grid, r - 1, c, island)
    _collect_island(grid, r + 1, c, island)
    _collect_island(grid, r, c - 1, island)
    _collect_island(grid, r, c + 1, island)
